litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1952,6 +2039,170 @@ protected LayoutState calculateLayoutState(
+    private final AtomicInteger runningThreadId = new AtomicInteger(-1);
+    private final ComponentContext context;
+    private final Component root;
+    private final int widthSpec;
+    private final int heightSpec;
+    private final boolean diffingEnabled;
+    @Nullable private final LayoutState previousLayoutState;
+    @Nullable private final TreeProps treeProps;
+    @Nullable private final String extraAttribution;
+    private volatile boolean isReleased = false;
+
+    private LayoutStateFuture(@Nullable final Object lock,
+        final ComponentContext context,
+        final Component root,
+        final int widthSpec,
+        final int heightSpec,
+        final boolean diffingEnabled,
+        @Nullable final LayoutState previousLayoutState,
+        @Nullable final TreeProps treeProps,
+        @CalculateLayoutSource final int source,
+        @Nullable final String extraAttribution) {
+      super(new Callable<LayoutState>() {
+        @Override
+        public LayoutState call() {
+          return calculateLayoutStateInternal(
+              lock,
+              context,
+              root,
+              widthSpec,
+              heightSpec,
+              diffingEnabled,
+              previousLayoutState,
+              treeProps,
+              source,
+              extraAttribution);
+        }
+      });
+
+      this.context = context;
+      this.root = root;
+      this.widthSpec = widthSpec;
+      this.heightSpec = heightSpec;
+      this.diffingEnabled = diffingEnabled;
+      this.previousLayoutState = previousLayoutState;
+      this.treeProps = treeProps;
+      this.extraAttribution = extraAttribution;
+    }
+
+    boolean isReleased() {
+      return isReleased;
+    }
+
+    @Override
+    public void run() {
+      if (runningThreadId.compareAndSet(-1, Process.myTid())) {
+        super.run();
+      }
+    }
+
+    @Nullable
+    @Override
+    public LayoutState get() {
+      final int runningThreadId = this.runningThreadId.get();
+      final int originalThreadPriority;
+      if (isMainThread() && !isDone() && runningThreadId != Process.myTid()) {
+        // Main thread is about to be blocked, raise the running thread priority.
+        originalThreadPriority = ThreadUtils
+            .tryRaiseThreadPriority(runningThreadId, Process.THREAD_PRIORITY_DISPLAY);
+      } else {
+        originalThreadPriority = THREAD_PRIORITY_INVALID;
+      }
+
+      final LayoutState layoutState;
+      try {
+        layoutState = super.get();
+      } catch (ExecutionException e) {
+        final Throwable cause = e.getCause();
+        if (cause instanceof RuntimeException) {
+          throw (RuntimeException) cause;
+        } else {
+          throw new RuntimeException(e.getMessage(), e);
+        }
+      } catch (InterruptedException | CancellationException e) {
+        throw new RuntimeException(e.getMessage(), e);
+      } finally {
+        if (originalThreadPriority != THREAD_PRIORITY_INVALID) {
```

Comment:
You can also maybe just use a boolean to keep track of whether to reset the priority instead of using a special value

Reply:
yeah, that's probably easier.